### ISSUE TITLE
Analysis of Classifier-Free Guidance Weight Schedulers

### ISSUE THREAD

**Issue by rootonchair** [huggingface/diffusers#7756]
### Model/Pipeline/Scheduler description

So there is an simple analysis and 1 line change in order to make Classifier-Free Guidance looks better
![cfg](https://github.com/huggingface/diffusers/assets/23548268/95e6801a-d734-4a50-bfbc-1d48a224f248)

I personally run some test to confirm
SD1.5 DDIM scheduler, 50 steps, "a photograph of an astronaut riding a horse"
guidance scale: 7.5
static
![sd_default](https://github.com/huggingface/diffusers/assets/23548268/884f5b8b-60fb-4b9d-b680-e6a3d6a047ab)

linear
![sd_default_linear](https://github.com/huggingface/diffusers/assets/23548268/7d3e70b4-7483-4857-a99a-5dbd522efc94)

guidance scale: 14.0
static
![sd_org](https://github.com/huggingface/diffusers/assets/23548268/f3059e37-c333-4383-a261-7563fe0f1b14)

linear
![sd_linear](https://github.com/huggingface/diffusers/assets/23548268/f0cad0e6-4e58-4c32-8394-cb588b6f95d0)

### Open source status

- [ ] The model implementation is available.
- [ ] The model weights are available (Only relevant if addition is not a scheduler).

### Provide useful links for the implementation

Paper: https://arxiv.org/abs/2404.13040

**Comment by rootonchair** [comment]
Update SDXL result
guidance scale: 14.0
static
![sdxl_org](https://github.com/huggingface/diffusers/assets/23548268/d1316ee7-21b6-4076-94a6-728f87581e20)
linear
![sdxl_org_linear](https://github.com/huggingface/diffusers/assets/23548268/4ed8dccc-5472-499f-8aee-11124937632c)

**Comment by DN6** [comment]
cc: @yiyixuxu and @asomoza for visibility.

**Comment by bghira** [comment]
what a neat way to make use of the knowledge already in the model!

**Comment by asomoza** [comment]
yes but there's a lot of techniques about manipulating the `CFG`, most of them without papers, I added the cutout one because I know it's really popular, makes the generations faster, and as a kind of example on how to manipulate the `CFG`

Maybe we should let the community add these ones later on? Other more popular ones are [automatic cfg](https://github.com/Extraltodeus/ComfyUI-AutomaticCFG) and [Dynamic Thresholding](https://github.com/mcmonkeyprojects/sd-dynamic-thresholding)

**Comment by YunhoKim21** [comment]
Hi, interested on your work. Is there an explanation of WHY this phenomena happens?

**Comment by bghira** [comment]
i'm not an expert on this, but on a cursory glance it seems to be basing the strength of guidance by the position in the timestep schedule. this also likely works because there's two types of attention being used by the model, with earlier timesteps being cross-attn (heavily relying on text conditional input) and later timesteps being self-attn (practically ignoring the prompt)

**Comment by sayakpaul** [comment]
@yiyixuxu @asomoza @a-r-r-o-w this could be a good candidate for modular diffusers no?

**Comment by bghira** [comment]
modular diffusers?

**Comment by asomoza** [comment]
> this could be a good candidate for modular diffusers no?

yes, actually all kind of variations the the CFG and also all the techniques that use the callbacks are the best initial candidates for the [modular diffusers](https://github.com/huggingface/diffusers/pull/9672).
